docs/src/manual/nonlinear.md

```diff
@@ -0,0 +1,567 @@
+```
+
+The arguments to [`@register`](@ref) are:
+
+ 1. The model in which the function is registered.
+ 2. A Julia symbol object which serves as the name of the user-defined function
+    in JuMP expressions. This name must not be the same as that of the function.
+ 3. The number of scalar input arguments that the function takes.
+ 4. A Julia method which computes the function.
+
+!!! warning
+    User-defined functions cannot be re-registered and will not update if you
+    modify the underlying Julia function. If you want to change a user-defined
+    function between solves, rebuild the model or use a different name.
+
+### Registered functions without macros
+
+The [`@register`](@ref) macro is syntactic sugar for the
+[`add_user_defined_function`](@ref) method. Thus, the non-macro version of the
+preceding example is:
+
+```@repl
+using JuMP
+square(x) = x^2
+f(x, y) = (x - 1)^2 + (y - 2)^2
+model = Model();
+my_square = add_user_defined_function(model, :my_square, 1, square)
+model[:my_square] = my_square
+my_f = add_user_defined_function(model, :my_f, 2, f)
+model[:my_f] = my_f
+@variable(model, x[1:2]);
+@objective(model, Min, my_f(x[1], my_square(x[2])))
+```
+
+This has two important consequences.
+
+First, you cannot register a user-defined function with the same name as an
+existing function. For example, a call to [`@register`](@ref) like:
+```julia
+julia> @register(model, square, 1, square)
+```
+will error because it is equivalent to:
+```julia
+julia> square = add_user_defined_function(model, :square, 1, square)
+ERROR: invalid redefinition of constant square
+Stacktrace:
+[...]
+```
+and `square` already exists as a Julia function.
+
+Second, you can obtain a reference to the user-defined function using the
+`model[:key]` syntax:
+
+```@repl
+using JuMP
+square(x) = x^2
+model = Model();
+@register(model, my_square, 1, square)
+my_square_2 = model[:my_square]
+```
+
+### Invalid redefinition of constant
+
+A common error encountered is `invalid redefinition of constant`. This occurs
+when the name of the user-defined function is the same as an existing function:
+```jldoctest nonlinear_invalid_redefinition
+julia> using JuMP
+
+julia> model = Model();
+
+julia> f(x) = x^2
+f (generic function with 1 method)
+
+julia> @register(model, f, 1, f)
+ERROR: invalid redefinition of constant f
+[...]
+```
+
+If you evaluate the function without registering it, JuMP will trace the
+function using operator overloading:
+```jldoctest nonlinear_invalid_redefinition
+julia> @variable(model, x);
+
+julia> f(x)
+x²
+```
+
+To force JuMP to treat `f` as a user-defined function and not trace it, register
+the function using [`add_user_defined_function`](@ref) and define a new method
+which manually creates a [`NonlinearExpr`](@ref):
+```jldoctest nonlinear_invalid_redefinition
+julia> _ = add_user_defined_function(model, :f, 1, f)
+UserDefinedFunction{typeof(f)}(:f, f)
+
+julia> f(x::AbstractJuMPScalar) = NonlinearExpr(:f, Any[x])
```

Comment:
I would also mention that you can use the returned value of `add_user_defined_function` and even overwrite the `f` function if in local scope.

Reply:
We mention the return value usage in `### Registered functions without macros`.

I don't know if we should tell people to overwrite the function in local scope. It's getting a bit niche, and there are some gotchas to be aware of.

Reply:
For the initial release I would try and keep the keep the docs as lean as possible and expand them as we understand the most common confusion that users encounter.

Reply:
There are also gotchas in the solution you suggest, it won't work if some arguments are JuMP scalars and some are JuMP expressions. It's also not so readable for non-expert. I feel like storing the return of the function in a variable like `nonlinear_f` and using that is much simpler. The expert user will know they can also use `f` id we are in local scope we don't even have to say it

Reply:
It's just documentation so it's not worth spending too much time on this I agree, we can wait and see